### PR TITLE
SIMSBIOHUB-569/589: Add Method Techniques and Manage Sampling Information Page

### PR DESCRIPTION
**Pull request by mauberti-bc**
## Links to Jira Tickets

https://apps.nrs.gov.bc.ca/int/jira/browse/SIMSBIOHUB-589
https://apps.nrs.gov.bc.ca/int/jira/browse/SIMSBIOHUB-569

## Description of Changes

- New page to manage both sampling sites and techniques.
  - Techniques are created under a survey, and are applied to a sampling method (as part of creating/editing sampling sites).

## Testing Notes

- CRUD for sampling sites
- CRUD for techniques

### CODE REVIEW

**Comment by MacDeluca** [comment]
1. Deleting techniques fails "rows was null or undefined, expected rows != null"
2. Deleting multiple techniques has no effect
3. In add sampling site page, add sampling method dialog renders with horizontal scroll on half view port
4. (Maybe out of scope) the sampling sites cards are rendering strangely

image

File: api/src/paths/codes.ts [hunk 1]
```diff
@@ -349,6 +350,25 @@ GET.apiDoc = {
                     }
                   }
                 }
+              },
+              attractants: {
```
Comment (MacDeluca) [inline api/src/paths/codes.ts:359]: Title / description specs would be helpful here

File: api/src/paths/codes.ts [hunk 2]
```diff
@@ -349,6 +350,25 @@ GET.apiDoc = {
...
+                type: 'array',
+                items: {
+                  type: 'object',
+                  additionalProperties: false,
+                  required: ['id', 'name', 'description'],
+                  properties: {
+                    id: {
```
Comment (MacDeluca) [inline api/src/paths/codes.ts:367]: minimum spec

File: api/src/paths/reference/get/technique-attribute.ts [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+                      name: {
+                        type: 'string'
+                      },
+                      description: {
+                        type: 'string',
+                        nullable: true
+                      },
+                      min: {
+                        type: 'integer',
+                        nullable: true
```
Comment (MacDeluca) [inline api/src/paths/reference/get/technique-attribute.ts:63]: minimum spec

File: api/src/paths/reference/get/technique-attribute.ts [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+                      },
+                      max: {
```
Comment (MacDeluca) [inline api/src/paths/reference/get/technique-attribute.ts:65]: minimum spec

File: api/src/paths/reference/get/technique-attribute.ts [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+      content: {
+        'application/json': {
+          schema: {
+            type: 'array',
+            items: {
+              type: 'object',
+              additionalProperties: false,
+              properties: {
+                method_lookup_id: {
+                  type: 'number'
```
Comment (MacDeluca) [inline api/src/paths/reference/get/technique-attribute.ts]: minimum spec

File: api/src/repositories/attractants-repository.ts [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+        throw new ApiExecuteSQLError('Failed to delete technique', [
+          'AttractantRepository->deleteTechnique',
+          'rows was null or undefined, expected rows != null'
+        ]);
+      }
+    }
+  }
+
+  /**
+   * Delete all technique attractants.
```
Comment (MacDeluca) [inline api/src/repositories/attractants-repository.ts]: 'Delete all technique attractants of a survey'

File: api/src/repositories/technique-attribute-repository.ts [hunk 7]
```diff
@@ -0,0 +1,529 @@
...
+  value: number;
+}
+
+export interface IQualitativeAttributePostData {
+  method_technique_attribute_qualitative_id?: number;
+  method_lookup_attribute_qualitative_option_id: string;
+  method_lookup_attribute_qualitative_id: string;
+}
+
+// export interface IQuantitativeAttributeRecord {
```
Comment (MacDeluca) [inline api/src/repositories/technique-attribute-repository.ts]: can we drop this dead code?💀

File: api/src/repositories/technique-attribute-repository.ts [hunk 8]
```diff
@@ -0,0 +1,529 @@
...
+const TechniqueAttributeQuantitative = z.object({
+  method_lookup_attribute_quantitative_id: z.string().uuid(),
+  name: z.string(),
+  description: z.string().nullable(),
+  unit: z.string().nullable(),
+  min: z.number().nullable(),
+  max: z.number().nullable()
+});
+
+// export interface ITechniqueAttributesObject {
```
Comment (MacDeluca) [inline api/src/repositories/technique-attribute-repository.ts]: 💀

File: api/src/repositories/technique-attribute-repository.ts [hunk 9]
```diff
@@ -0,0 +1,529 @@
...
+});
+
+const TechniqueAttributeQualitative = z.object({
+  method_lookup_attribute_qualitative_id: z.string().uuid(),
+  name: z.string(),
+  description: z.string().nullable(),
+  options: z.array(TechniqueAttributeQualitativeOption)
+});
+
+// export interface IGetTechniqueAttributes {
```
Comment (MacDeluca) [inline api/src/repositories/technique-attribute-repository.ts]: 💀

File: api/src/paths/project/{projectId}/survey/{surveyId}/sample-site/{surveySampleSiteId}/sample-method/index.ts [hunk 10]
```diff
@@ -94,7 +94,7 @@ GET.apiDoc = {
                     survey_sample_site_id: {
                       type: 'integer'
                     },
-                    method_lookup_id: {
+                    method_technique_id: {
                       type: 'integer'
```
Comment (MacDeluca) [inline api/src/paths/project/{projectId}/survey/{surveyId}/sample-site/{surveySampleSiteId}/sample-method/index.ts]: min

File: api/src/services/technique-service.ts [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+  ITechniqueRowDataForInsert,
+  ITechniqueRowDataForUpdate,
+  TechniqueObject,
+  TechniqueRepository
+} from '../repositories/technique-repository';
+import { ApiPaginationOptions } from '../zod-schema/pagination';
+import { AttractantService } from './attractants-service';
+import { DBService } from './db-service';
+
+export class TechniqueService extends DBService {
```
Comment (MacDeluca) [inline api/src/services/technique-service.ts:21]: jsdoc

File: app/src/assets/images/observations-overlay.png
Comment (MacDeluca) [inline app/src/assets/images/observations-overlay.png:1]: 🐻

File: app/src/components/dialog/ErrorDialog.tsx [hunk 12]
```diff
@@ -93,21 +96,25 @@ export const ErrorDialog = (props: IErrorDialogProps) => {
...
+            <Box>
+              <Link color="primary" onClick={() => setIsExpanded(!isExpanded)}>
+                {isExpanded ? 'Hide detailed error message' : 'Show detailed error message'}
+              </Link>
+              <Collapse in={isExpanded}>
+                <ErrorDetailsList errors={props.dialogErrorDetails} />
+              </Collapse>
+            </Box>
+          ) : (
+            <></>
```
Comment (MacDeluca) [inline app/src/components/dialog/ErrorDialog.tsx]: Does this need to be a fragment for the stack? Or does null work?

File: app/src/components/fields/AutocompleteField.tsx [hunk 13]
```diff
@@ -66,41 +73,80 @@ const AutocompleteField = <T extends string | number>(props: IAutocompleteField<
...
+          </Box>
+        );
+      }}
+      renderInput={(params) => {
+        return (
+          <TextField
+            {...params}
+            required={props.required}
+            label={props.label}
+            InputLabelProps={
```
Comment (MacDeluca) [inline app/src/components/fields/AutocompleteField.tsx]: Remove prop?

File: app/src/components/fields/DateTimeFields.tsx [hunk 14]
```diff
@@ -75,8 +91,8 @@ export const DateTimeFields = <FormikPropsType,>(props: IDateTimeFieldsProps<For
                 name: dateName,
                 required: dateRequired,
                 variant: 'outlined',
-                error: get(touched, dateName) && Boolean(get(errors, dateName)),
-                helperText: get(touched, dateName) && get(errors, dateName),
+                error: get(touched, dateName) && (dateError || Boolean(get(errors, dateName))),
+                helperText: get(touched, dateName) && (dateHelperText || get(errors, dateName)),
```
Comment (MacDeluca) [inline app/src/components/fields/DateTimeFields.tsx:95]: Should we flip 'dateHelperText' and 'get(errors, dateName)' so the error will render if we are providing the dateHelperText?

```js
helperText: get(touched, dateName) && (get(errors, dateName) || dateHelperText)
```

File: app/src/components/fields/DateTimeFields.tsx [hunk 15]
```diff
@@ -120,8 +136,8 @@ export const DateTimeFields = <FormikPropsType,>(props: IDateTimeFieldsProps<For
                 name: timeName,
                 required: timeRequired,
                 variant: 'outlined',
-                error: get(touched, timeName) && Boolean(get(errors, timeName)),
-                helperText: get(touched, timeName) && get(errors, timeName),
+                error: get(touched, timeName) && (timeError || Boolean(get(errors, timeName))),
+                helperText: get(touched, timeName) && (timeHelperText || get(errors, timeName)),
```
Comment (MacDeluca) [inline app/src/components/fields/DateTimeFields.tsx:140]: same here

File: app/src/features/surveys/sampling-information/manage/SamplingSiteManageHeader.tsx [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+import { Link as RouterLink } from 'react-router-dom';
+
+export interface SamplingSiteManageHeaderProps {
+  project_id: number;
+  project_name: string;
+  survey_id: number;
+  survey_name: string;
+}
+
+export const SamplingSiteManageHeader = (props: SamplingSiteManageHeaderProps) => {
```
Comment (MacDeluca) [inline app/src/features/surveys/sampling-information/manage/SamplingSiteManageHeader.tsx:20]: jsdoc

File: app/src/features/surveys/sampling-information/manage/SamplingSiteManagePage.tsx [hunk 17]
```diff
@@ -0,0 +1,32 @@
+import Container from '@mui/material/Container';
+import Paper from '@mui/material/Paper';
+import Stack from '@mui/material/Stack';
+import { SamplingSiteManageHeader } from 'features/surveys/sampling-information/manage/SamplingSiteManageHeader';
+import { SamplingSiteManageSiteList } from 'features/surveys/sampling-information/sites/manage/SamplingSiteManageSiteList';
+import { SamplingTechniqueContainer } from 'features/surveys/sampling-information/techniques/SamplingTechniqueContainer';
+import { useProjectContext, useSurveyContext } from 'hooks/useContext';
+
+export const SamplingSiteManagePage = () => {
```
Comment (MacDeluca) [inline app/src/features/surveys/sampling-information/manage/SamplingSiteManagePage.tsx:14]: jsdoc

File: app/src/features/surveys/sampling-information/methods/components/SamplingMethodForm.tsx [hunk 18]
```diff
@@ -0,0 +1,143 @@
...
+            label="Technique"
+            name="technique.method_technique_id"
+            options={
+              techniques?.map((option) => ({
+                value: option.method_technique_id,
+                label: option.name,
+                subText: option.description ?? undefined
+              })) ?? []
+            }></SelectWithSubtextField>
+          {/* <AutocompleteField
```
Comment (MacDeluca) [inline app/src/features/surveys/sampling-information/methods/components/SamplingMethodForm.tsx]: 💀

File: api/src/paths/codes.ts (hunk 1 above)
Comment (NickPhura) [reply api/src/paths/codes.ts:359]: Added some stuff, though I probably need to go through and do the same for all of the codes, maybe as tech debt separately.

File: api/src/paths/codes.ts (hunk 2 above)
Comment (NickPhura) [reply api/src/paths/codes.ts:367]: added

File: api/src/paths/reference/get/technique-attribute.ts (hunk 3 above)
Comment (NickPhura) [reply api/src/paths/reference/get/technique-attribute.ts:63]: negative values might need to be allowed

File: api/src/paths/reference/get/technique-attribute.ts (hunk 5 above)
Comment (NickPhura) [reply api/src/paths/reference/get/technique-attribute.ts]: added

File: api/src/repositories/attractants-repository.ts (hunk 6 above)
Comment (NickPhura) [reply api/src/repositories/attractants-repository.ts]: updated

File: api/src/repositories/technique-attribute-repository.ts (hunk 7 above)
Comment (NickPhura) [reply api/src/repositories/technique-attribute-repository.ts]: Yes! I noticed there was some redundancy before, but must have forgotten to remove the old interfaces. Fixed

File: api/src/repositories/technique-attribute-repository.ts (hunk 9 above)
Comment (NickPhura) [reply api/src/repositories/technique-attribute-repository.ts]: removed

File: api/src/paths/project/{projectId}/survey/{surveyId}/sample-site/{surveySampleSiteId}/sample-method/index.ts (hunk 10 above)
Comment (NickPhura) [reply api/src/paths/project/{projectId}/survey/{surveyId}/sample-site/{surveySampleSiteId}/sample-method/index.ts]: Updated, added some other missing fields to the spec as well.

File: api/src/services/technique-service.ts (hunk 11 above)
Comment (NickPhura) [reply api/src/services/technique-service.ts:21]: Updated

File: app/src/assets/images/observations-overlay.png
Comment (NickPhura) [reply app/src/assets/images/observations-overlay.png:1]: 🤷‍♂️

File: app/src/components/dialog/ErrorDialog.tsx (hunk 12 above)
Comment (NickPhura) [reply app/src/components/dialog/ErrorDialog.tsx]: changed to `null`

File: app/src/components/fields/AutocompleteField.tsx (hunk 13 above)
Comment (NickPhura) [reply app/src/components/fields/AutocompleteField.tsx]: yep, removed

File: app/src/components/fields/DateTimeFields.tsx (hunk 14 above)
Comment (NickPhura) [reply app/src/components/fields/DateTimeFields.tsx:95]: We actually never use dateHelperText anywhere, so I might just remove it

File: app/src/components/fields/DateTimeFields.tsx (hunk 15 above)
Comment (NickPhura) [reply app/src/components/fields/DateTimeFields.tsx:140]: We actually never use dateHelperText anywhere, so I might just remove it

File: app/src/features/surveys/sampling-information/manage/SamplingSiteManagePage.tsx (hunk 17 above)
Comment (NickPhura) [reply app/src/features/surveys/sampling-information/manage/SamplingSiteManagePage.tsx:14]: Added

File: app/src/features/surveys/sampling-information/methods/components/SamplingMethodForm.tsx (hunk 18 above)
Comment (NickPhura) [reply app/src/features/surveys/sampling-information/methods/components/SamplingMethodForm.tsx]: Purged.

File: api/src/repositories/technique-attribute-repository.ts (hunk 8 above)
Comment (NickPhura) [reply api/src/repositories/technique-attribute-repository.ts]: Removed

File: app/src/features/surveys/sampling-information/manage/SamplingSiteManageHeader.tsx (hunk 16 above)
Comment (NickPhura) [reply app/src/features/surveys/sampling-information/manage/SamplingSiteManageHeader.tsx:20]: added

**Comment by NickPhura** [comment]
> 1. Deleting techniques fails "rows was null or undefined, expected rows != null"
> 
>     2. Deleting multiple techniques has no effect
> 
>     3. In add sampling site page, add sampling method dialog renders with horizontal scroll on half view port
> 
>     4. (Maybe out of scope) the sampling sites cards are rendering strangely
> 
> 

@MacQSL (1, 2, and 3) should be fixed. I didnt address 4. Will probably leave that for a separate UI polish ticket.

**Review by MacDeluca — APPROVED** [review]
😶‍🌫️

**Pull request merged**